net/net.go

```diff
@@ -174,6 +179,73 @@ func NewNetwork(
 	return t, nil
 }
 
+func (n *net) countRecords(ctx context.Context, tid thread.ID, rid cid.Cid) (int64, error) {
+	var (
+		cursor  = rid
+		counter int64 = 0
+	)
+	sk, err := n.store.ServiceKey(tid)
+	if err != nil {
+		return 0, err
+	}
+
+	for cursor.Defined() {
+		r, err := cbor.GetRecord(ctx, n, cursor, sk)
+		if err != nil {
+			return 0, err
+		}
+		cursor = r.PrevID()
+		counter += 1
+	}
+	return counter, nil
+}
+
+func (n *net) migrateHeadsIfNeeded(ctx context.Context, ls lstore.Logstore) (err error) {
+	threadIds, err := ls.Threads()
+	if err != nil {
+		return err
+	}
+
+	log.Info("Starting migrating heads")
```

Comment:
nit: this is going to log "Starting migration heads" even if we don't end up doing a migration, which might be confusing? What about just a) changing this to "Checking for head migrations" / "Finished checking for head migrations" or b) using a flag outside the loop to check if any migrations are required, and only lazily logging if true?